fetch/README.md

```diff
@@ -0,0 +1,134 @@
+
+## Table of Contents
+
+- [Overview](#overview)
+- [Spec Proposal Info](#spec-proposal-info)
+- [Wire protocol](#wire-protocol)
+- [Interfaces](#interfaces)
+- [Implementation Details](#implementation-details)
+- [Removing existing relay protocol](#removing-existing-relay-protocol)
+
+## Overview
+
+The fetch protocol is a means for a libp2p node to get data from another based on a given key. It generally fulfills
+the contract:
+
+`Fetch(key) (value, statusCode)`
+
+## Spec Proposal Info
+### Context
+
+Currently there is no standard protocol for doing simple data retrieval. Sometimes a small data retrieval protocol is all
+that's needed to get the job done. This is especially true when it comes to augmenting or layering protocols together.
+
+### Motivation
+
+This specification is helpful because the protocol is already part of a proposal to add a persistence layer on top of
+[PubSub](../pubsub).
+
+There have been numerous discussion revolving around how to allow PubSub to be persistent, some of which go back multiple
+years (more background information is available [here](https://github.com/libp2p/go-libp2p/issues/698)).
+Initial suggestions included modifying or leveraging the existing PubSub spec and implementations.
+However, it was decided that PubSub should remain agnostic to any sort of persistence,
+and that persistence should be layered on top of PubSub as a separate protocol.
+
+There is currently an [IPNS-over-PubSub spec PR](https://github.com/ipfs/specs/pull/218) that describes how the Fetch
+protocol can be used to create a persistence layer on top of PubSub and therefore greatly improve IPNS performance.
+
+### Scope
+
+This is one small protocol that as of right now is only being used to improve the
+[go-libp2p-pubsub-router](https://github.com/libp2p/go-libp2p-pubsub-router). This protocol should not require any one
+who doesn't want to use it to change anything about their current use.
+
+### Potential Effects
+
+1. A persistent pubsub protocol easily built on top of Fetch
+2. A reusable fetch protocol that can be used to augment various other protocols
+(e.g. add the ability to directly ask peers for data to a DHT)
+3. Might be abused/overused given the simple nature of the protocol
+
+### Expected Feature Set and Tentative Technical Directions
+ 
+Should support: `Fetch(key) (value, statusCode)`
+
+However, the level of specificity in the types of the above variables has wiggle room if people are interested.
+The `go-libp2p-pubsub-router` implementation requires:
+ 
+ `key`: At least as generic as a UTF-8 string
+ 
+ `value`: At least as generic as a byte array
+ 
+ `statusCode`: Supports at least `SUCCESS` and `NOT_FOUND`
+
+## Wire protocol
+
+The libp2p protocol ID for this protocol is `/libp2p/fetch/0.0.1`
+
+### Message Format
+The messages in the Fetch protocol use on of the following protobufs (proto3 syntax):
+
+```
+message RequestLatest {
```

Comment:
`RequestLatest` doesn't rhyme with the fetch protocol, we should call the request just `Request` and the response `Response`.

Reply:
Or `FetchRequest` and `FetchResponse`.